File: Code/controllers/supervisor/simulator_distance.cpp
#define RAND ((float) rand()/RAND_MAX)
#include <cmath>
#include <Point2d.h>
#include <supervisor.cc>
#define TOTAL_ITER 100

const Point2d wall_1_left_pt = Point2d(-0.625,0.0);
const Point2d wall_1_right_pt = Point2d(-0.25,0.0);
const Point2d wall_2_upper_pt = Point2d(0.125,0.625);
const Point2d wall_2_lower_pt = Point2d(0.125,-0.225);

double average_distance = 0.0;

for(int i=0; i<TOTAL_ITER; i++){
    double travel_distance;
    Point2d robot_rand = rand_coord();
    Point2d event_rand = rand_coord();
    bool intersects_wall_1 = get_line_intersection(robot_rand.x,robot_rand.y,event_rand.x,event_rand.y,-0.625,0.0,-0.25,0.0);
    bool intersects_wall_2 = get_line_intersection(robot_rand.x,robot_rand.y,event_rand.x,event_rand.y,0.125,0.625,0.125,-0.225);
    if(intersects_wall_1){
        travel_distance+=robot_rand.Distance(wall_1_right_pt);
        if(intersects_wall_2){
            travel_distance+=wall_1_right_pt.Distance(wall_2_lower_pt)+wall_2_lower_pt.Distance(event_rand);
        }else{
            travel_distance+=wall_1_right_pt.Distance(event_rand);
        }
    }else if(intersects_wall_2){
        travel_distance+=robot_rand.Distance(wall_2_lower_pt)+wall_2_lower_pt.Distance(event_rand);
    }else{
        travel_distance+=robot_rand.Distance(event_rand);
    }
    average_distance+=travel_distance;
}
average_distance=average_distance/TOTAL_ITER;
printf("Average distance traveled (assuming only presence of 1 robot and 1 event) is %f", average_distance);

File: simulator_distance.py
import random as rand
import math

# Constants
TOTAL_ITER = 10000

# Wall endpoints
wall_1_left_pt = (-0.625, 0.0)
wall_1_right_pt = (-0.25, 0.0)
wall_2_upper_pt = (0.125, 0.625)
wall_2_lower_pt = (0.125, -0.225)

def distance(p1, p2):
    """Calculate Euclidean distance between two points."""
    return math.sqrt((p2[0] - p1[0])**2 + (p2[1] - p1[1])**2)

def ccw(A, B, C):
    """Check if three points are listed in counter-clockwise order."""
    return (C[1] - A[1]) * (B[0] - A[0]) > (B[1] - A[1]) * (C[0] - A[0])

def segments_intersect(A, B, C, D):
    """Check if line segment AB intersects line segment CD."""
    return (ccw(A, C, D) != ccw(B, C, D)) and (ccw(A, B, C) != ccw(A, B, D))

def get_line_intersection(robot_x, robot_y, event_x, event_y, wall_x1, wall_y1, wall_x2, wall_y2):
    """Check if the segment between robot and event intersects with a wall segment."""
    A = (robot_x, robot_y)
    B = (event_x, event_y)
    C = (wall_x1, wall_y1)
    D = (wall_x2, wall_y2)
    return segments_intersect(A, B, C, D)

def rand_coord():
    """Generate random coordinates with the arena constraints."""
    rand_x = rand.uniform(-0.625, 0.625)
    rand_y = rand.uniform(-0.625, 0.625)
    while((rand_x <= -0.20 and (rand_y < 0.055 or rand_y > -0.055))or((rand_x > 0.07 and rand_x < 0.18)and(rand_y >= -0.275)) ):
        rand_x = rand.uniform(-0.625, 0.625)
        rand_y = rand.uniform(-0.625, 0.625)
    return (rand_x, rand_y)

def main():
    average_distance = 0.0
    
    for i in range(TOTAL_ITER):
        travel_distance = 0.0
        
        # Generate random positions
        robot_rand = rand_coord()
        event_rand = rand_coord()
        
        # Check for intersections with walls
        intersects_wall_1 = get_line_intersection(
            robot_rand[0], robot_rand[1], 
            event_rand[0], event_rand[1],
            wall_1_left_pt[0], wall_1_left_pt[1],
            wall_1_right_pt[0], wall_1_right_pt[1]
        )
        
        intersects_wall_2 = get_line_intersection(
            robot_rand[0], robot_rand[1],
            event_rand[0], event_rand[1],
            wall_2_upper_pt[0], wall_2_upper_pt[1],
            wall_2_lower_pt[0], wall_2_lower_pt[1]
        )
        
        # Calculate travel distance based on wall intersections
        if intersects_wall_1:
            travel_distance += distance(robot_rand, wall_1_right_pt)
            if intersects_wall_2:
                travel_distance += distance(wall_1_right_pt, wall_2_lower_pt)
                travel_distance += distance(wall_2_lower_pt, event_rand)
            else:
                travel_distance += distance(wall_1_right_pt, event_rand)
        elif intersects_wall_2:
            travel_distance += distance(robot_rand, wall_2_lower_pt)
            travel_distance += distance(wall_2_lower_pt, event_rand)
        else:
            travel_distance += distance(robot_rand, event_rand)
        
        average_distance += travel_distance
    
    average_distance /= TOTAL_ITER
    print("Average distance traveled (assuming only presence of 1 robot and 1 event) is {:.6f}".format(average_distance))
    
    return average_distance

if __name__ == "__main__":
    main()
